walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,84 @@
+  */  
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      Collection<TimeRange> result;
+      Collection<String> requestAttendees = request.getAttendees();
+      Collection<String> requestOptionalAttendees = request.getOptionalAttendees();
+      
+      ArrayList<TimeRange> resultList = new ArrayList<TimeRange>();
+      ArrayList<TimeRange> resultWithOptional = new ArrayList<TimeRange>();
+      
+      int previousEventEndTime = TimeRange.START_OF_DAY;
+      int previousEventTimeOptionalAttendees = TimeRange.START_OF_DAY;
+      long requestedTime = request.getDuration();
+
+      result = Arrays.asList();
+      
+      if(TimeRange.END_OF_DAY < requestedTime){
+            return result;
+      }
+
+      if(events.size() == 0){
+          result = Arrays.asList(TimeRange.WHOLE_DAY);
+          return result;
+      }
+      
+
+     for (Event event : events) {
+         Collection<String> eventAttendees = event.getAttendees();
+         TimeRange eventInterval = event.getWhen();
+         int eventStartTime = eventInterval.start();
+         if(requestAttendees.containsAll(eventAttendees)){
+            if(previousEventEndTime + requestedTime <= eventStartTime){
+                    resultList.add(TimeRange.fromStartEnd(previousEventEndTime, eventStartTime, false));
+            }
+            if(previousEventEndTime <  eventInterval.end())
+                previousEventEndTime =  eventInterval.end();
+
+            if(previousEventTimeOptionalAttendees + requestedTime <= eventStartTime){
+                resultWithOptional.add(TimeRange.fromStartEnd(previousEventTimeOptionalAttendees, eventStartTime, false));
+                }
+            if(previousEventTimeOptionalAttendees <  eventInterval.end())
+                previousEventTimeOptionalAttendees =  eventInterval.end();
+            } 
+
+         if(requestOptionalAttendees.containsAll(eventAttendees)){
+            if(previousEventTimeOptionalAttendees + requestedTime <= eventStartTime){
+                    resultWithOptional.add(TimeRange.fromStartEnd(previousEventTimeOptionalAttendees, eventStartTime, false));
+                }
+            if(previousEventTimeOptionalAttendees <  eventInterval.end())
+                    previousEventTimeOptionalAttendees =  eventInterval.end();      
+            }
+        }
+
+     if(previousEventEndTime <= TimeRange.END_OF_DAY)
+        resultList.add(TimeRange.fromStartEnd(previousEventEndTime, TimeRange.END_OF_DAY, true));
+     
+     if(previousEventTimeOptionalAttendees < TimeRange.END_OF_DAY)
+            resultWithOptional.add(TimeRange.fromStartEnd(previousEventTimeOptionalAttendees, TimeRange.END_OF_DAY, true));
+
+     if(resultWithOptional.size() == 0)
+        result = resultList;
+     else
+        result = resultWithOptional; 
+
+      return result;
```

Comment:
Overall, this looks good. I feel that the code could have benefit from some comments, this will make it easier for others, and yourself in a month from now, to quickly grasp the flow and conditions.